File: repositories/booking_repo.py
from __future__ import annotations

from typing import Any

from bson import ObjectId
from pymongo import ASCENDING, DESCENDING
from pymongo import ReturnDocument

from core.database import db
from schemas.booking import BookingCreate, BookingOut

_BOOKING_INDEXES_READY = False


async def _ensure_booking_indexes() -> None:
    global _BOOKING_INDEXES_READY
    if _BOOKING_INDEXES_READY:
        return

    await db.bookings.create_index("customer_id", name="idx_booking_customer_id")
    await db.bookings.create_index("cleaner_id", name="idx_booking_cleaner_id")
    await db.bookings.create_index("status", name="idx_booking_status")
    await db.bookings.create_index("schedule", name="idx_booking_schedule")
    await db.bookings.create_index("place_id", name="idx_booking_place_id")
    await db.bookings.create_index(
        "payment_id",
        name="idx_booking_payment_id_unique",
        unique=True,
        sparse=True,
    )
    _BOOKING_INDEXES_READY = True


def _id_filter(booking_id: str) -> dict:
    if ObjectId.is_valid(booking_id):
        return {"_id": ObjectId(booking_id)}
    return {"_id": booking_id}


async def create_booking(payload: BookingCreate) -> BookingOut:
    await _ensure_booking_indexes()
    result = await db.bookings.insert_one(payload.model_dump(mode="json"))
    stored = await db.bookings.find_one({"_id": result.inserted_id})
    return BookingOut(**stored) # type: ignore


async def get_booking_by_id(booking_id: str) -> BookingOut | None:
    await _ensure_booking_indexes()
    row = await db.bookings.find_one(_id_filter(booking_id))
    if row is None:
        return None
    return BookingOut(**row)


async def get_bookings(
    *,
    filter_dict: dict | None = None,
    start: int = 0,
    stop: int = 100,
) -> list[BookingOut]:
    await _ensure_booking_indexes()
    query = filter_dict or {}
    cursor = db.bookings.find(query).skip(start).limit(max(0, stop - start))
    items: list[BookingOut] = []
    async for row in cursor:
        items.append(BookingOut(**row))
    return items


async def get_bookings_history(
    *,
    filter_dict: dict[str, Any] | None = None,
    scope: str,
    now_epoch: int,
    payment_status: str | None,
    cursor_offset: int,
    page_size: int,
    scheduled_sort: str,
) -> list[BookingOut]:
    await _ensure_booking_indexes()
    query: dict[str, Any] = dict(filter_dict or {})

    if scope in {"upcoming", "current"}:
        query["schedule"] = {"$gte": now_epoch}
    elif scope in {"past", "history"}:
        query["schedule"] = {"$lt": now_epoch}

    direction = ASCENDING if scheduled_sort == "asc" else DESCENDING
    pipeline: list[dict[str, Any]] = [
        {"$match": query},
        {"$addFields": {"_booking_id_str": {"$toString": "$_id"}}},
        {
            "$lookup": {
                "from": "payment_transactions",
                "localField": "_booking_id_str",
                "foreignField": "booking_id",
                "as": "payment_tx",
            }
        },
        {
            "$addFields": {
                "_payment_status_effective": {
                    "$let": {
                        "vars": {"tx": {"$arrayElemAt": ["$payment_tx", 0]}},
                        "in": {"$ifNull": [{"$toLower": "$$tx.status"}, "pending"]},
                    }
                }
            }
        },
    ]
    if payment_status is not None:
        pipeline.append({"$match": {"_payment_status_effective": payment_status}})

    pipeline.extend(
        [
            {"$sort": {"schedule": direction, "_id": direction}},
            {"$skip": cursor_offset},
            {"$limit": page_size + 1},
            {
                "$project": {
                    "_booking_id_str": 0,
                    "payment_tx": 0,
                    "_payment_status_effective": 0,
                }
            },
        ]
    )

    items: list[BookingOut] = []
    cursor = db.bookings.aggregate(pipeline)
    async for row in cursor:
        items.append(BookingOut(**row))
    return items


async def update_booking_fields(booking_id: str, update_dict: dict) -> BookingOut | None:
    await _ensure_booking_indexes()
    if not update_dict:
        return await get_booking_by_id(booking_id)
    row = await db.bookings.find_one_and_update(
        _id_filter(booking_id),
        {"$set": update_dict},
        return_document=ReturnDocument.AFTER,
    )
    if row is None:
        return None
    return BookingOut(**row)


async def transition_booking_status(
    *,
    booking_id: str,
    expected_status: str,
    update_dict: dict,
) -> BookingOut | None:
    await _ensure_booking_indexes()
    row = await db.bookings.find_one_and_update(
        {**_id_filter(booking_id), "status": expected_status},
        {"$set": update_dict},
        return_document=ReturnDocument.AFTER,
    )
    if row is None:
        return None
    return BookingOut(**row)


async def delete_booking(booking_id: str):
    await _ensure_booking_indexes()
    return await db.bookings.delete_one(_id_filter(booking_id))

File: app/server/repositories/booking-repo.ts
import type { Collection, Filter, Sort } from 'mongodb'
import { getDb } from '@/server/core/mongo'
import {
  BookingOut,
  type BookingDoc,
  type BookingOut as BookingOutType,
  type BookingStatus,
  type BookingPaymentStatus,
  type BookingHistoryScope,
} from '@/server/schemas/booking'
import { idFilter, fromDoc } from './_helpers'

/**
 * Data access for the `bookings` collection.
 * Ported from `repositories/booking_repo.py`. Only this layer touches Mongo.
 *
 * Indexes (per docs/migration/02-data-model.md): customer_id, cleaner_id,
 * status, schedule, place_id, and a unique sparse payment_id.
 */

let indexesReady = false

function collection(): Collection<BookingDoc> {
  return getDb().collection<BookingDoc>('bookings')
}

export async function ensureIndexes(): Promise<void> {
  if (indexesReady) return
  const col = collection()
  await col.createIndex({ customer_id: 1 }, { name: 'idx_booking_customer_id' })
  await col.createIndex({ cleaner_id: 1 }, { name: 'idx_booking_cleaner_id' })
  await col.createIndex({ status: 1 }, { name: 'idx_booking_status' })
  await col.createIndex({ schedule: 1 }, { name: 'idx_booking_schedule' })
  await col.createIndex({ place_id: 1 }, { name: 'idx_booking_place_id' })
  await col.createIndex(
    { payment_id: 1 },
    { name: 'idx_booking_payment_id', unique: true, sparse: true },
  )
  indexesReady = true
}

function parse(doc: unknown): BookingOutType {
  return BookingOut.parse(fromDoc(doc))
}

export async function createBooking(doc: BookingDoc): Promise<BookingOutType> {
  await ensureIndexes()
  const result = await collection().insertOne(doc)
  const stored = await collection().findOne(idFilter(String(result.insertedId)))
  return parse(stored)
}

export async function getBookingById(id: string): Promise<BookingOutType | null> {
  await ensureIndexes()
  const row = await collection().findOne(idFilter(id))
  return row ? parse(row) : null
}

export interface BookingFilter {
  customerId?: string
  cleanerId?: string
  status?: BookingStatus
  paymentStatus?: BookingPaymentStatus
}

function buildFilter(f: BookingFilter): Filter<BookingDoc> {
  const query: Filter<BookingDoc> = {}
  if (f.customerId) query.customer_id = f.customerId
  if (f.cleanerId) query.cleaner_id = f.cleanerId
  if (f.status) query.status = f.status
  if (f.paymentStatus) query.payment_status = f.paymentStatus
  return query
}

/** Simple filtered fetch (scheduled ascending), no pagination. */
export async function getBookings(f: BookingFilter): Promise<BookingOutType[]> {
  await ensureIndexes()
  const rows = await collection().find(buildFilter(f)).sort({ schedule: 1 }).toArray()
  return rows.map(parse)
}

export interface BookingHistoryQuery extends BookingFilter {
  scope?: BookingHistoryScope
  scheduledSort?: 'asc' | 'desc'
  /** Cursor = the `_id` of the last item from the previous page. */
  cursor?: string
  pageSize?: number
  /** Reference "now" (epoch seconds) for scope time filtering. */
  now?: number
}

export interface BookingHistoryResult {
  items: BookingOutType[]
  nextCursor: string | null
  pageSize: number
}

/**
 * Cursor/offset paginated history with scope time filtering and scheduled sort.
 * `scope` filters by `schedule` relative to `now` (upcoming = future, past =
 * already elapsed). Cursor pagination is `_id`-based and stable within a sort.
 */
export async function getBookingsHistory(q: BookingHistoryQuery): Promise<BookingHistoryResult> {
  await ensureIndexes()

  const pageSize = q.pageSize && q.pageSize > 0 ? q.pageSize : 20
  const sortDir = q.scheduledSort === 'asc' ? 1 : -1
  const now = q.now ?? Math.floor(Date.now() / 1000)

  const query = buildFilter(q) as Filter<BookingDoc> & Record<string, unknown>

  // Scope time filter on the scheduled epoch.
  if (q.scope === 'upcoming') query.schedule = { $gte: now }
  else if (q.scope === 'past') query.schedule = { $lt: now }

  // Cursor: continue after the last seen _id, honoring the sort direction.
  if (q.cursor) {
    const { toObjectId } = await import('./_helpers')
    const cursorOp = sortDir === 1 ? '$gt' : '$lt'
    query._id = { [cursorOp]: toObjectId(q.cursor) }
  }

  const sort: Sort = { schedule: sortDir, _id: sortDir }

  // Fetch one extra to detect whether a further page exists.
  const rows = await collection()
    .find(query)
    .sort(sort)
    .limit(pageSize + 1)
    .toArray()

  const hasMore = rows.length > pageSize
  const page = hasMore ? rows.slice(0, pageSize) : rows
  const nextCursor = hasMore ? String(page[page.length - 1]?._id) : null

  return { items: page.map(parse), nextCursor, pageSize }
}

export async function updateBooking(
  id: string,
  set: Partial<BookingDoc>,
): Promise<BookingOutType | null> {
  await ensureIndexes()
  await collection().updateOne(idFilter(id), { $set: set })
  return getBookingById(id)
}

/** Count bookings for a cleaner (optionally filtered by status). Derivation source for jobsDone. */
export async function countForCleaner(cleaner_id: string, status?: BookingStatus): Promise<number> {
  await ensureIndexes()
  const query: Filter<BookingDoc> = { cleaner_id }
  if (status) query.status = status
  return collection().countDocuments(query)
}

/**
 * Cleaner job feed: jobs assigned to this cleaner PLUS the unassigned PENDING
 * pool, excluding jobs this cleaner has declined. Scheduled ascending.
 */
export async function getCleanerJobFeed(cleanerId: string): Promise<BookingOutType[]> {
  await ensureIndexes()
  const query: Filter<BookingDoc> & Record<string, unknown> = {
    $or: [{ cleaner_id: cleanerId }, { cleaner_id: null, status: 'PENDING' }],
    declinedBy: { $ne: cleanerId },
  }
  const rows = await collection().find(query).sort({ schedule: 1 }).toArray()
  return rows.map(parse)
}

/** Record that a cleaner has passed on an (unassigned) job. */
export async function addDecline(bookingId: string, cleanerId: string): Promise<void> {
  await ensureIndexes()
  await collection().updateOne(idFilter(bookingId), {
    $addToSet: { declinedBy: cleanerId },
    $set: { lastUpdated: Math.floor(Date.now() / 1000) },
  })
}
